examples/model_selection/plot_grid_search_digits.py

```diff
@@ -13,56 +13,167 @@
+# select the best candidate from the `cv_results_` returns from the
+# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate selected,
+# it will be automatically refitted by the
+# :class:`~sklearn.model_selection.GridSearchCV` instance. Here, the strategy
+# is to short list the model which are the best in terms of precision and
+# recall. From the selected models, we finally select the fastest model at
+# predicting.
+#
+# Note that these choices are completely arbitrary.
+import pandas as pd
+
+
+def refit_strategy(cv_results):
+    """Define the strategy to select the best estimator.
+
+    The strategy defined here is to filter-out all results below a precision threshold
+    of 0.98, rank the remaining by recall and keep all models with one standard
+    deviation of the best by recall. Once these models are selected, we can select the
+    fastest model to predict.
+
+    Parameters
+    ----------
+    cv_results : dict of numpy (masked) ndarrays
+        CV results as returned by the `GridSearchCV`.
+
+    Returns
+    -------
+    best_index : int
+        The index of the best estimator as it appears in `cv_results`.
+    """
+    # print the info about the grid-search for the different scores
+    precision_threshold = 0.98
+    for score in scores:
+        print(f"\nGrid scores on development set using {score}:\n")
+        for mean, std, params in zip(
+            cv_results[f"mean_test_{score}"],
+            cv_results[f"std_test_{score}"],
+            cv_results["params"],
+        ):
+            print(f"{mean:0.3f} (+/-{std * 2:0.03f}) for {params}")
+
+    cv_results_ = pd.DataFrame(cv_results)
+
+    # Filter-out all results below the threshold
+    cv_results_ = cv_results_[
+        cv_results_["mean_test_precision_macro"] > precision_threshold
+    ]
+
+    cv_results_ = cv_results_[
+        [
+            "mean_score_time",
+            "mean_test_recall_macro",
+            "mean_test_precision_macro",
+            "rank_test_recall_macro",
+            "rank_test_precision_macro",
+        ]
+    ]
+
+    # Select the most performant models in terms of recall
+    # (within 1 sigma from the best)
+    mask_best_recall = cv_results_["rank_test_recall_macro"] == 1
+    best_recall_std = cv_results_.std()["mean_test_recall_macro"]
+    best_recall_index = cv_results_.loc[
+        mask_best_recall, "mean_test_recall_macro"
+    ].idxmin()
+    best_recall = cv_results_.loc[best_recall_index, "mean_test_recall_macro"]
+    best_recall_threshold = best_recall - best_recall_std
+
+    cv_results_ = cv_results_[
+        cv_results_["mean_test_recall_macro"] > best_recall_threshold
+    ]
+    print(
+        "The following models are the most performance in terms of "
+        f"recall:\n\n {cv_results_}"
+    )
+
+    # From the best candidates, select the fastest model to predict
+    best_index = cv_results_.idxmin(axis="rows")["mean_score_time"]
```

Comment:
`best_index` could renamed to `fastest_top_recall_high_precision_index` or similar to be more descriptive.